service/history/task/task.go

```diff
@@ -224,6 +242,7 @@ func (t *taskImpl) HandleErr(
 				t.scope.RecordTimer(metrics.TaskAttemptTimerPerDomain, time.Duration(t.attempt))
 				t.logger.Error("Critical error processing task, retrying.",
 					tag.Error(err), tag.OperationCritical, tag.TaskType(t.GetTaskType()))
+				t.ReportCorruptWorkflowToWatchDog()
```

Comment:
If the watchdog isn't able to repair in a couple of retries, its chances of succeeding later are too slim, so we should back off from retrying this too many times in order to free up the workflow and the DB. Calling this for each attempt after the `criticalRetryCount()` is too much. Perhaps run it for every N multiples of the `criticalRetryCount` to retry more than once but less than a reasonable limit.